src/awx_plugins/credentials/aws_assumerole.py

```diff
@@ -0,0 +1,128 @@
+                                  },
+                                 {'id': 'role_arn',
+                                  'label': 'AWS ARN Role Name',
+                                  'type': 'string',
+                                  'secret': True,
+                                  'help_text': _('The ARN Role Name to be assumed in AWS'),
+                                  },
+                                 ],
+                      'metadata': [{'id': 'identifier',
+                                    'label': 'Identifier',
+                                    'type': 'string',
+                                    'help_text': _('The name of the key in the assumed AWS role to fetch [AccessKeyId | SecretAccessKey | SessionToken].'),
+                                    },
+                                   ],
+                      'required': ['role_arn'],
+                      }
+
+
+def aws_assumerole_getcreds(access_key, secret_key, role_arn, external_id):
+    if (access_key is None or len(access_key) == 0) and (
+            secret_key is None or len(secret_key) == 0
+    ):
+        # Connect using credentials in the EE
+        connection = boto3.client(service_name='sts')
+    else:
+        # Connect to AWS using provided credentials
+        connection = boto3.client(
+            service_name='sts',
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key,
+        )
+    try:
+        response = connection.assume_role(
+            RoleArn=role_arn,
+            RoleSessionName='AAP_AWS_Role_Session1',
+            ExternalId=external_id,
+        )
+    except ClientError as ce:
+        raise ValueError(f'Got a bad client response from AWS: {ce.msg}.')
+
+    credentials = response.get('Credentials', {})
+
+    return credentials
+
+
+def aws_assumerole_backend(**kwargs):
+    """This backend function actually contacts AWS to assume a given role for
+    the specified user."""
+    access_key = kwargs.get('access_key')
+    secret_key = kwargs.get('secret_key')
+    role_arn = kwargs.get('role_arn')
+    external_id = kwargs.get('external_id')
+    identifier = kwargs.get('identifier')
+
+    # Generate a unique SHA256 hash for combo of user access key and ARN
+    # This should allow two users requesting the same ARN role to have
+    # separate credentials, and should allow the same user to request
+    # multiple roles.
+    #
+    credential_key_hash = hashlib.sha256(
+        (str(access_key or '') + role_arn).encode('utf-8'),
+    )
+    credential_key = credential_key_hash.hexdigest()
+
+    credentials = _aws_cred_cache.get(credential_key, None)
+
+    # If there are no credentials for this user/ARN *or* the credentials
+    # we have in the cache have expired, then we need to contact AWS again.
+    #
+    if (credentials is None) or (
+        credentials['Expiration'] < datetime.datetime.now(
+            credentials['Expiration'].tzinfo,
+        )
+    ):
```

Comment:
Readability tip: labeling parts of conditionals with named variables eliminates the need for using code comments for story-telling.
```suggestion
    creds_expire_at = credentials.get('Expiration')
    creds_expired = (
        creds_expire_at and
        creds_expire_at < datetime.now(credentials['Expiration'].tzinfo)
    )
    if creds_expired:
```
(this suggestion requires `from datetime import datetime` at the top, which would also simplify a number of places by dropping those confusing `datetime.datetime.` references)

Reply:
can you explain this a bit further? my evaluation of this is that cred_expire_at is being declared before use. try to hack together something simple in a python environment and I got a NameError
```
>>> b = (c:= 1 and c)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'c' is not defined
```

I think I understand where you are coming from a readability standpoint vs code comments.

Reply:
@thedoubl3j you misplaced the paren and so did I, it seems. The assignment expression needs to be wrapped:
```pycon
$ python3.13t
Python 3.13.0rc1 experimental free-threading build (main, Aug 13 2024, 04:25:17) [GCC 13.3.1 20240614] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> b = (c:= 1 and c)
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    b = (c:= 1 and c)
                   ^
NameError: name 'c' is not defined
>>> b = (c:= 1) and c
>>> b
1
>>> c
1
```

I've updated the above suggested comment, and it should be ready to be accepted by clicking a button on the GH UI.

Reply:
ahh, that is what I wasn't seeing. cheers good sir. added.